kernel/src/process/process_vm/init_stack/mod.rs

```diff
@@ -366,17 +363,27 @@ fn generate_random_for_aux_vec() -> [u8; 16] {
 }
 
 /// A reader to parse the content of an `InitStack`.
-pub struct InitStackReader {
+pub struct InitStackReader<'a> {
     base: Vaddr,
+    vm_space: &'a Arc<VmSpace>,
     /// The mapping address of the `InitStack`.
     map_addr: usize,
 }
 
-impl InitStackReader {
+impl InitStackReader<'_> {
     /// Reads argc from the process init stack
     pub fn argc(&self) -> Result<u64> {
         let stack_base = self.init_stack_bottom();
-        get_current_userspace!().read_val(stack_base)
+        let page_base_addr = stack_base.align_down(PAGE_SIZE);
+
+        let mut cursor = self
+            .vm_space
+            .cursor(&(page_base_addr..page_base_addr + PAGE_SIZE))?;
+        if let VmItem::Mapped { frame, .. } = cursor.query()? {
+            Ok(frame.read_val::<u64>(stack_base - page_base_addr)?)
+        } else {
```

Comment:
This is a general (and also well known) advice on writing readable code: **do not treat the correct and error branches as equal; return early on error to avoid unnecessary indentation levels.** This is so wide accepted practice that the Rust language assists you with the `let-else` construct.

```rust
let VmItem::Mapped { frame, .. } = cursor.query()? else {
    return_errno_with_message!(Errno::EACCES, "Page not accessible");
};

// Continue on with the correct, main logic...
```

This is a general technique to improve code readability. Use it as much as you can. While this function can benefit from this technique, the improvement on readability is greater on more complex code. For example, apply this technique to the `argv` and `envp` methods (which are much more complex) and see the results for yourself 😄

Reply:
improved with `let-else`